src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -596,82 +700,199 @@ namespace nbl::ext::imgui
+			
+				if (unallocatedSize != 0u)
 				{
-					if (!vBinding.memory->map({ 0ull, vBinding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
-						logger->log("Could not map device memory for vertex buffer data!", system::ILogger::ELL_WARNING);
+					// retry, second attempt cull frees and execute deferred memory deallocation of offsets no longer in use
+					unallocatedSize = m_mdi.streamingTDBufferST->multi_allocate(timeout, MDI_ALLOCATION_COUNT, multiAllocParams.offsets.data(), multiAllocParams.byteSizes.data(), MDI_ALIGNMENTS.data());
 
-					assert(vBinding.memory->isCurrentlyMapped());
-				}
+					if (unallocatedSize != 0u)
+					{
+						logger->log("Could not multi alloc mdi buffer!", system::ILogger::ELL_ERROR);
 
-				{
-					if (!iBinding.memory->map({ 0ull, iBinding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
-						logger->log("Could not map device memory for index buffer data!", system::ILogger::ELL_WARNING);
-				
-					assert(iBinding.memory->isCurrentlyMapped());
+						auto getOffsetStr = [&](const MDI::COMPOSE_T::value_type offset) -> std::string
+						{
+							return offset == MDI::COMPOSE_T::invalid_value ? "invalid_value" : std::to_string(offset);
+						};
+
+						logger->log("[mdi streaming buffer] = \"%s\" bytes", system::ILogger::ELL_ERROR, std::to_string(mdiBuffer->getSize()).c_str());
+						logger->log("[requested] = \"%s\" bytes", system::ILogger::ELL_ERROR, std::to_string(mdiBufferByteSize).c_str());
+						logger->log("[unallocated] = \"%s\" bytes", system::ILogger::ELL_ERROR, std::to_string(unallocatedSize).c_str());
+
+						logger->log("[MDI::EBC_DRAW_INDIRECT_STRUCTURES offset] = \"%s\" bytes", system::ILogger::ELL_ERROR, getOffsetStr(multiAllocParams.offsets[MDI::EBC_DRAW_INDIRECT_STRUCTURES]).c_str());
+						logger->log("[MDI::EBC_ELEMENT_STRUCTURES offset] = \"%s\" bytes", system::ILogger::ELL_ERROR, getOffsetStr(multiAllocParams.offsets[MDI::EBC_ELEMENT_STRUCTURES]).c_str());
+						logger->log("[MDI::EBC_INDEX_BUFFERS offset] = \"%s\" bytes", system::ILogger::ELL_ERROR, getOffsetStr(multiAllocParams.offsets[MDI::EBC_INDEX_BUFFERS]).c_str());
+						logger->log("[MDI::EBC_VERTEX_BUFFERS offset] = \"%s\" bytes", system::ILogger::ELL_ERROR, getOffsetStr(multiAllocParams.offsets[MDI::EBC_VERTEX_BUFFERS]).c_str());
+
+						exit(0x45); // TODO: handle OOB memory requests, probably need to extend the mdi buffer/let user pass more size at init
+					}
 				}
+			}
+
+			const uint32_t drawCount = multiAllocParams.byteSizes[MDI::EBC_DRAW_INDIRECT_STRUCTURES] / sizeof(VkDrawIndexedIndirectCommand);
+			{
+				auto binding = mdiBuffer->getBoundMemory();
+
+				if(!binding.memory->isCurrentlyMapped())
+					if (!binding.memory->map({ 0ull, binding.memory->getAllocationSize() }, IDeviceMemoryAllocation::EMCAF_READ))
+						logger->log("Could not map device memory!", system::ILogger::ELL_WARNING);
```

Comment:
don't handle this at runtime, check during `create` factory method / constructor

Reply:
resolved